scipy/interpolate/_interpolate.py

```diff
@@ -1009,6 +1013,9 @@ def __call__(self, x, nu=0, extrapolate=None):
             x = self.x[0] + (x - self.x[0]) % (self.x[-1] - self.x[0])
             extrapolate = False
 
+        with self._c_lock:
+            pass
+
```

Comment:
This is an optimistic-lock; it expects that multiple threads to be calling non-mutating functions most of the time and allowing them to run concurrently. It behaves like a turnstile, allowing threads that arrived first to pass, and then blocks if another thread calls `extend`, until it finishes updating the polynomial coefficients. 

While this solution may yield dirty-reads, it would allow for better concurrency, by not serializing each call that depends on the coefficients. Instead, the concurrency management and ordering is left to the user.

Reply:
Not sure I understand this comment, and most certainly a comment is needed not only on GH but also in code.

Also, if it's something to do with the `extend` method, can the lock moved to extend? <s>If not, why it's only in `__call__` but not in other functions (`.derivative`, `.integrate` etc)</s>

Reply:
This one still confuses me, I've to admit. Let me try rephrasing my confusion:

1. This with-stanza acquires and releases the lock, so that if two threads arrive simultaneously, one arrives to this line and the other is doing something else, that other one is stopped and waits until the lock is released -- -is this correct? If it is, why is it needed, why is it useful and how does this helps anything?

2. The very need for this is because of existence of `.extend` method which mutates the class state, is this correct? If correct,cannot one lock the `.extend` internals instead?

3. What is this pattern, is it something known/standard or is it something you conjured out of thin air?

4. Can you actually show a scenario where this is needed? And explain how it works in this scenario?
Previous discussion seems to suggest something rather esoteric, with threaded use of reads and `.extend` --- a concrete scenario would really help.

Reply:
As I mentioned in the first comment, this lock acts like a "turnstile with a block", imagine n persons entering a wall through a turnstile, once they are in, they are independent of each other, despite entering in a serialized fashion. Now, imagine there's a service mode that the turnstile can enter in which it blocks while it goes maintenance, during this time no people can enter the wall, until the service mode gets released.

This is the same scenario posed here, what's behind the wall is the code that can be run by each thread independent of each other, assuming that the polynomial coefficients haven't changed at all. What the so-called "service mode" in the analogy does is changing the coefficients while stopping everyone else from accessing them.

This is more of a thought experiment that can be an exotic scenario in the practice, that's why I posed the question about signaling in the docs that calls to extend should not be interleaved with concurrent calls to any of the interpolation subclasses that inherit from `_PPolyBase`, since it is not thread-safe.

Reply:
Okay, the turnstile analogy I get, thank you. That the implementation looks bizarre can be attributed to me not being used to this, so can discount that :-). 

What I'm not sure about is the status of this: 

- this is a best-effort trick, to remedy _some_ of free-threading usages, correct?
- (if yes) is there a way to estimate how exotic should a user load be for this trick to not work?

Let's consider an example: https://docs.scipy.org/doc/scipy/tutorial/interpolate/extrapolation_examples.html#cubicspline-extend-the-boundary-conditions

- if we keep land this PR as is, how would the example need changing?
- if we land the PR without this turnstile, how would the example need changing?

Reply:
This mutation goes against the guidelines declared on https://github.com/Quansight-Labs/free-threaded-compatibility?tab=readme-ov-file#suggested-plan-of-attack, also SciPy should not anticipate exotic use cases like the one that initially motivated this whole "optimistic lock" approach.

The resolution in this case was to document the `extend` function and mark it as not thread-safe when used with interleaved interpolation methods.